encoder/vaapiencoder_vp9.cpp

```diff
@@ -0,0 +1,323 @@
+#include <algorithm>
+
+namespace YamiMediaCodec {
+
+// each frame can have at most 3 reference frames, golden, alt and last
+enum maxSizeValues {
+    kMaxReferenceFrames = 8,
+    kMaxWidth = 8192,
+    kMaxHeight = kMaxWidth
+};
+
+enum VP9QPValues { kMinQPValue = 9, kDefaultQPValue = 60, kMaxQPValue = 127 };
+
+enum frameType { kKeyFrame, kInterFrame };
+
+enum levelValues { kDefaultSharpnessLevel, kDefaultFilterLevel = 10 };
+
+class VaapiEncPictureVP9 : public VaapiEncPicture {
+public:
+    VaapiEncPictureVP9(const ContextPtr& context, const SurfacePtr& surface,
+                       int64_t timeStamp)
+        : VaapiEncPicture(context, surface, timeStamp)
+    {
+        return;
+    }
+
+    VAGenericID getCodedBufferID() { return m_codedBuffer->getID(); }
+};
+
+VaapiEncoderVP9::VaapiEncoderVP9()
+    : m_frameCount(0)
+{
+    m_videoParamCommon.profile = VAProfileVP9Profile0;
+    m_videoParamCommon.rcParams.minQP = kMinQPValue;
+    m_videoParamCommon.rcParams.maxQP = kMaxQPValue;
+    m_videoParamCommon.rcParams.initQP = kDefaultQPValue;
+
+    //add extra surfaces to operate due to kMaxReferenceFrames
+    //vaapi_encoder class will create 5 extra surfaces already
+    m_maxOutputBuffer = kMaxReferenceFrames;
+}
+
+VaapiEncoderVP9::~VaapiEncoderVP9() {}
+
+YamiStatus VaapiEncoderVP9::getMaxOutSize(uint32_t* maxSize)
+{
+    FUNC_ENTER();
+    *maxSize = m_maxCodedbufSize;
+    return YAMI_SUCCESS;
+}
+
+void VaapiEncoderVP9::resetParams()
+{
+    m_maxCodedbufSize = width() * height() * 3 / 2;
+
+    // adding extra padding.  In particular small resolutions require more
+    // space depending on other quantization parameters during execution. The
+    // value below is a good compromise
+    m_maxCodedbufSize += 0x1000;
+}
+
+YamiStatus VaapiEncoderVP9::start()
+{
+    FUNC_ENTER();
+    resetParams();
+    return VaapiEncoderBase::start();
+}
+
+void VaapiEncoderVP9::flush()
+{
+    FUNC_ENTER();
+    m_frameCount = 0;
+    m_reference.clear();
+    VaapiEncoderBase::flush();
+}
+
+YamiStatus VaapiEncoderVP9::stop()
+{
+    flush();
+    return VaapiEncoderBase::stop();
+}
+
+YamiStatus VaapiEncoderVP9::setParameters(VideoParamConfigType type,
+                                          Yami_PTR videoEncParams)
+{
+    YamiStatus status = YAMI_SUCCESS;
+    FUNC_ENTER();
+    if (!videoEncParams)
+        return YAMI_INVALID_PARAM;
+
+    switch (type) {
+    default:
+        status = VaapiEncoderBase::setParameters(type, videoEncParams);
+        break;
+    }
+    return status;
+}
+
+YamiStatus VaapiEncoderVP9::getParameters(VideoParamConfigType type,
+                                          Yami_PTR videoEncParams)
+{
+    FUNC_ENTER();
+    if (!videoEncParams)
+        return YAMI_INVALID_PARAM;
+
+    return VaapiEncoderBase::getParameters(type, videoEncParams);
+}
+
+YamiStatus VaapiEncoderVP9::doEncode(const SurfacePtr& surface,
+                                     uint64_t timeStamp, bool forceKeyFrame)
```

Comment:
It not cover by our test application. Maybe we can write a unit test to enforce this.